onmt/utils/alignment.py

```diff
@@ -120,43 +120,85 @@ def to_word_align(src, tgt, subword_align, m_src='joiner', m_tgt='joiner'):
     return " ".join(word_align)
 
 
+# Helper functions
+def begin_uppercase(token):
+    return token == "｟mrk_begin_case_region_U｠"
+
+
+def end_uppercase(token):
+    return token == "｟mrk_end_case_region_U｠"
+
+
+def begin_case(token):
+    return token == "｟mrk_case_modifier_C｠"
+
+
+def case_markup(token):
+    return token in [
+        "｟mrk_begin_case_region_U｠",
+        "｟mrk_end_case_region_U｠",
+        "｟mrk_case_modifier_C｠"]
+
+
 def subword_map_by_joiner(subwords,
-                          marker=SubwordMarker.JOINER,
-                          case_markup=SubwordMarker.CASE_MARKUP):
+                          original_subwords=None,
+                          marker=SubwordMarker.JOINER):
     """Return word id for each subword token (annotate by joiner)."""
+
     flags = [1] * len(subwords)
+    j = 0
+    finished = False
     for i, tok in enumerate(subwords):
-        if (tok.endswith(marker) or
-                (tok in case_markup and tok.find("end") < 0)):
+
+        previous_tok = subwords[i-1] if i else ""
+        previous_tok_2 = subwords[i-2] if i > 1 else ""
+        current_original_subword = "" if not original_subwords \
+            else original_subwords[j] if j < len(original_subwords) else ""
+
+        if tok.startswith(marker) and tok != current_original_subword:
+            flags[i] = 0
+        elif (previous_tok.endswith(marker)
+                or begin_case(previous_tok)
+                or begin_uppercase(previous_tok)) \
+                and not finished:
+            flags[i] = 0
+        elif previous_tok_2.endswith(marker) \
+                and case_markup(previous_tok) \
+                and not finished:
+            flags[i] = 0
+        elif end_uppercase(tok) and tok != current_original_subword:
             flags[i] = 0
-        if (tok.startswith(marker) or
-                (tok in case_markup and tok.find("end") >= 0)):
-            assert i >= 1 and flags[i-1] != 0, \
-                "Sentence `{}` not correct!".format(" ".join(subwords))
-            flags[i-1] = 0
-    word_group = list(accumulate([0] + flags[:-1]))
+        else:
+            finished = False
+            if tok == current_original_subword:
+                finished = True
+            j += 1
+
+    flags[0] = 0
+    word_group = list(accumulate(flags))
+
+    if original_subwords:
+        assert max(word_group) < len(original_subwords)
     return word_group
```

Comment:
This function is not very readable.
`previous_tok` // `previous_tok_2` ?
At least add some comments to explain the few steps of logic here.